File: view/android/VerificationCodeJavaDemo/app/src/main/java/com/example/verificationcodejavademo/widget/DragImageView.java
package com.example.verificationcodejavademo.widget;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.verificationcodejavademo.R;
import com.example.verificationcodejavademo.utils.DisplayUtil;

/**
 * Date:2020/5/19
 * author:wuyan
 */
public class DragImageView extends FrameLayout implements SeekBar.OnSeekBarChangeListener {

    private SeekBar drag_sb;
    private FrameLayout drag_fl_content;
    private ImageView drag_iv_cover;
    private ImageView drag_iv_block;
    private View drag_v_flash;
    private DiyStyleTextView drag_tv_tips;
    private TextView drag_tv_tips2;
    private Bitmap cover;
    private Bitmap block;

    private Long animeTime = 333L;
    private float timeUse = 0.0f;
    private Long timeTemp = 0l;

    private Handler handler = new Handler();

    public DragImageView(@NonNull Context context) {
        super(context);
        init();
    }

    public DragImageView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public DragImageView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        View.inflate(getContext(), R.layout.drag_view, this);
        drag_sb = findViewById(R.id.drag_sb);
        drag_sb.setOnSeekBarChangeListener(this);
        drag_fl_content = findViewById(R.id.drag_fl_content);
        drag_iv_cover = findViewById(R.id.drag_iv_cover);
        drag_iv_block = findViewById(R.id.drag_iv_block);
        drag_v_flash = findViewById(R.id.drag_v_flash);
        drag_tv_tips = findViewById(R.id.drag_tv_tips);
        drag_tv_tips2 = findViewById(R.id.drag_tv_tips2);
        drag_sb.setMax(getContext().getResources().getDisplayMetrics().widthPixels);
        reset();
    }

    //重置
    public void reset() {
        int position = drag_sb.getProgress();
        if (position != 0) {
            ValueAnimator animator = ValueAnimator.ofFloat(1f, 0.0f);
            animator.setDuration(animeTime).start();
            animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator animation) {
                    float f = (float) animation.getAnimatedValue();
                    drag_sb.setProgress((int) (position * f));
                }
            });
        }
        tipsShowAnime(false);
        tips2ShowAnime(true);
        setSbThumb(R.drawable.drag_btn_n);
        drag_sb.setEnabled(true);
        drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress));
        drag_iv_block.setVisibility(VISIBLE);
    }

    /**
     * 设置资源
     *
     * @param cover 拼图
     * @param block 滑块
     */
    public void setUp(Bitmap cover, Bitmap block) {
        this.cover = cover;
        this.block = block;

        int w = cover.getWidth();
        int h = cover.getHeight();

        LayoutParams l = (LayoutParams) drag_iv_cover.getLayoutParams();
        l.width = DisplayUtil.dip2px(getContext(), (float) w);
        l.height = DisplayUtil.dip2px(getContext(), (float) h);

        drag_iv_cover.setLayoutParams(l);
        drag_iv_cover.setImageBitmap(cover);

        LayoutParams layoutParams2 = (LayoutParams) drag_iv_block.getLayoutParams();
        layoutParams2.width = DisplayUtil.dip2px(getContext(), (float) block.getWidth());
        layoutParams2.height = DisplayUtil.dip2px(getContext(), (float) block.getHeight());
        drag_iv_block.setLayoutParams(layoutParams2);
        drag_iv_block.setImageBitmap(block);

        setLocation(1f * cover.getWidth() / cover.getHeight(), cover.getWidth());

    }

    //设置容器宽高
    private void setLocation(float cover_wph, int cover_w) {
        int w = cover_w;
        int h = (int) (w / cover_wph);
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) drag_fl_content.getLayoutParams();
        layoutParams.width = DisplayUtil.dip2px(getContext(), (float) w);
        layoutParams.height = DisplayUtil.dip2px(getContext(), (float) h);
        drag_fl_content.setLayoutParams(layoutParams);
    }

    //验证成功:滑块隐藏，成功提示语，滑块不能滑动，设置setSbThumb，SeekBar的背景
    public void ok() {
        blockHideAnime();

        int penset;
        if (timeUse > 1) {
            penset = (int) (99 - (timeUse - 1) / 0.1f);
        } else {
            penset = (int) (99 - 0 / 0.1f);
        }
        if (penset < 1) penset = 1;
        drag_tv_tips.setText(String.format("拼图成功: 耗时%.1f秒,打败了%d%%的用户!", timeUse, penset));
        tipsShowAnime(true);

        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                tipsShowAnime(false);
                tips2ShowAnime(true);
                drag_sb.setEnabled(true);

                int position = drag_sb.getProgress();
                ValueAnimator animator = ValueAnimator.ofFloat(1.0f, 0.0f);
                animator.setDuration(animeTime).start();
                animator.addUpdateListener(animation -> {
                    float f = (float) animation.getAnimatedValue();
                    drag_sb.setProgress((int) (position * f));
                    setSbThumb(R.drawable.drag_btn_n);
                    drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress));
                });

            }
        },1500);

        drag_sb.setEnabled(false);
        setSbThumb(R.drawable.drag_btn_success);
        drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress_success));

    }

    //成功完成拼图滑块消失
    private void blockHideAnime() {
        AlphaAnimation alphaAnimation = new AlphaAnimation(1f, 0f);
        alphaAnimation.setDuration(animeTime);
        drag_iv_block.setAnimation(alphaAnimation);
        drag_iv_block.setVisibility(GONE);
    }

    //验证失败
    public void fail() {
        drag_tv_tips.setText("拼图失败: 请重新拖曳滑块到正确的位置!");
        tipsShowAnime(true);
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                tipsShowAnime(false);
                tips2ShowAnime(true);
                drag_sb.setEnabled(true);

                int position = drag_sb.getProgress();
                ValueAnimator animator = ValueAnimator.ofFloat(1.0f, 0.0f);
                animator.setDuration(animeTime).start();
                animator.addUpdateListener(animation -> {
                    float f = (float) animation.getAnimatedValue();
                    drag_sb.setProgress((int) (position * f));
                    setSbThumb(R.drawable.drag_btn_n);
                    drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress));
                });

            }
        },1500);
        drag_sb.setEnabled(false);
        setSbThumb(R.drawable.drag_btn_error);
        drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress_fail));
    }

    //弹出提示语是否显示
    private void tipsShowAnime(boolean isShow) {
        TranslateAnimation translateAnimation = new TranslateAnimation(Animation.RELATIVE_TO_SELF, 0f, Animation.RELATIVE_TO_SELF, 0f, Animation.RELATIVE_TO_SELF, isShow ? 1f : 0f, Animation.RELATIVE_TO_SELF, isShow ? 0f : 1f);
        translateAnimation.setDuration(animeTime);
        drag_tv_tips.setAnimation(translateAnimation);
        drag_tv_tips.setVisibility(isShow ? VISIBLE : GONE);
    }

    //seekBar上面的提示语是否显示
    private void tips2ShowAnime(boolean isShow) {
        AlphaAnimation alphaAnimation = new AlphaAnimation(isShow ? 0f : 1f, isShow ? 1f : 0f);
        alphaAnimation.setDuration(animeTime);
        drag_tv_tips2.setAnimation(alphaAnimation);
        drag_tv_tips2.setVisibility(isShow ? VISIBLE : GONE);
    }


    //===================seekbar监听===================

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        int cw = drag_iv_cover.getMeasuredWidth();
        int bw = drag_iv_block.getMeasuredWidth();
        ViewGroup.MarginLayoutParams l = (MarginLayoutParams) drag_iv_block.getLayoutParams();
        l.leftMargin = (cw - bw) * progress / seekBar.getMax();
        drag_iv_block.setLayoutParams(l);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        setSbThumb(R.drawable.drag_btn_n);
        drag_sb.setProgressDrawable(getResources().getDrawable(R.drawable.drag_seek_progress));
        drag_iv_block.setVisibility(VISIBLE);
        drag_iv_cover.setImageBitmap(cover);
        tips2ShowAnime(false);
        timeTemp = System.currentTimeMillis();
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        timeUse = (System.currentTimeMillis() - timeTemp) / 1000f;//滑动滑块所花的时间，单位s
        if (dragListenner != null) {
            dragListenner.onDrag(DisplayUtil.px2dip(
                    getContext(), (drag_iv_cover.getMeasuredWidth() - drag_iv_block.getMeasuredWidth()) * 1f * seekBar.getProgress() / seekBar.getMax()
            ));
        }


    }

    //设置seekBar的Thumb
    private void setSbThumb(int id) {
        Drawable drawable = getResources().getDrawable(id);
        drawable.setBounds(drag_sb.getThumb().getBounds());
        drag_sb.setThumb(drawable);
        drag_sb.setThumbOffset(0);
    }

    //设置seekBar是否可以滑动
    public void setSBUnMove(boolean isMove) {
        drag_sb.setEnabled(isMove);
    }

    //设置滑动监听
    private DragListenner dragListenner;

    interface DragListenner {
        void onDrag(double position);
    }

    public void setDragListenner(DragListenner dragListenner) {
        this.dragListenner = dragListenner;
    }
}

File: view/android/VerificationCodeDemo/app/src/main/java/com/example/verificationcodedemo/widget/DragImageView.kt
package com.example.verificationcodedemo.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Bitmap
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.TranslateAnimation
import android.widget.FrameLayout
import android.widget.SeekBar
import com.example.verificationcodedemo.R
import com.example.verificationcodedemo.utils.DisplayUtil
import com.example.verificationcodedemo.utils.DisplayUtil.px2dip
import kotlinx.android.synthetic.main.drag_view.view.*

/**
 * Date:2020/5/6
 * author:wuyan
 */
class DragImageView : FrameLayout, SeekBar.OnSeekBarChangeListener {

    private val showTipsTime = 1500
    private val animeTime = 333
    private val flashTime = 800

    private var sb: SeekBar? = null
    private var cover: Bitmap? = null
    private var block: Bitmap? = null

    //===================seekbar监听===================
    private var timeTemp: Long = 0
    private var timeUse: Float = 0.toFloat()

    //失败延时重置控件
    private val resetRun = Runnable {
        tipsShowAnime(false)
        tips2ShowAnime(true)
        sb!!.isEnabled = true

        val position = sb!!.progress
        val animator = ValueAnimator.ofFloat(1f, 0.0f)
        animator.setDuration(animeTime.toLong()).start()
        animator.addUpdateListener { animation ->
            val f = animation.animatedValue as Float
            sb!!.progress = (position * f).toInt()
            setSbThumb(R.drawable.drag_btn_n)
            sb!!.progressDrawable = context.resources.getDrawable(R.drawable.drag_seek_progress)
        }
    }

    //监听
    private var dragListenner: DragListenner? = null

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init()
    }

    private fun init() {
        View.inflate(context, R.layout.drag_view, this)
        //设置需要字体颜色为红色的内容
        drag_tv_tips!!.setColorRegex("拼图|成功|失败|正确|[\\d\\.%]+", -0x8aeaf)
        sb = findViewById(R.id.drag_sb)
        sb!!.max = context.resources.displayMetrics.widthPixels
        sb!!.setOnSeekBarChangeListener(this)
        reset()
    }


    /**
     * 设置资源
     *
     * @param cover 拼图
     * @param block 滑块
     */
    fun setUp(cover: Bitmap, block: Bitmap) {
        this.cover = cover
        this.block = block

        val w = cover.width
        val h = cover.height
        val l = drag_iv_cover!!.layoutParams
        l.width = DisplayUtil.dip2px(context, w.toFloat())
        l.height = DisplayUtil.dip2px(context, h.toFloat())
        drag_iv_cover!!.layoutParams = l
        drag_iv_cover!!.setImageBitmap(cover)

        val l2 = drag_iv_block!!.layoutParams as ViewGroup.MarginLayoutParams
        l2.height = DisplayUtil.dip2px(context, block.height.toFloat())
        l2.width = DisplayUtil.dip2px(context, block.width.toFloat())
        drag_iv_block!!.layoutParams = l2
        drag_iv_block!!.setImageBitmap(block)
        setLocation(1f * cover.width / cover.height, cover.width)
    }

    /**
     * 设置比例大小
     *
     * @param cover_wph  图片bili
     * @param block_size 滑块大小占高比
     */
    private fun setLocation(cover_wph: Float, cover_w: Int) {
        post {
            val w = cover_w
            val h = (w / cover_wph).toInt()
            val l = drag_fl_content!!.layoutParams
            l.width = DisplayUtil.dip2px(context, w.toFloat())
            l.height = DisplayUtil.dip2px(context, h.toFloat())
            drag_fl_content!!.layoutParams = l
        }

    }

    fun ok() {
        blockHideAnime()
        var penset: Int
        if (timeUse > 1) {
            penset = (99 - (timeUse - 1) / 0.1f).toInt()
        } else {
            penset = (99 - 0 / 0.1f).toInt()
        }
        if (penset < 1) penset = 1
        drag_tv_tips!!.text = String.format("拼图成功: 耗时%.1f秒,打败了%d%%的用户!", timeUse, penset)
        tipsShowAnime(true)
        flashShowAnime()
        sb!!.isEnabled = false
        setSbThumb(R.drawable.drag_btn_success)
        sb!!.progressDrawable = context.resources.getDrawable(R.drawable.drag_seek_progress_success)
    }

    fun fail() {
        drag_tv_tips!!.text = "拼图失败: 请重新拖曳滑块到正确的位置!"
        tipsShowAnime(true)
        handler.postDelayed(resetRun, showTipsTime.toLong())
        sb!!.isEnabled = false
        setSbThumb(R.drawable.drag_btn_error)
        sb!!.progressDrawable = context.resources.getDrawable(R.drawable.drag_seek_progress_fail)
    }

    fun reset() {
        val position = sb!!.progress
        if (position != 0) {
            val animator = ValueAnimator.ofFloat(1f, 0.0f)
            animator.setDuration(animeTime.toLong()).start()
            animator.addUpdateListener { animation ->
                val f = animation.animatedValue as Float
                sb!!.progress = (position * f).toInt()
            }
        }
        tipsShowAnime(false)
        tips2ShowAnime(true)
        sb!!.isEnabled = true
        drag_v_flash!!.visibility = View.GONE
        setSbThumb(R.drawable.drag_btn_n)
        sb!!.progressDrawable = context.resources.getDrawable(R.drawable.drag_seek_progress)
        drag_iv_block!!.visibility = View.VISIBLE
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        val cw = drag_iv_cover!!.measuredWidth
        val bw = drag_iv_block!!.measuredWidth
        val l = drag_iv_block!!.layoutParams as ViewGroup.MarginLayoutParams
        l.leftMargin = (cw - bw) * progress / seekBar.max
        drag_iv_block!!.layoutParams = l
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
        setSbThumb(R.drawable.drag_btn)
        sb!!.progressDrawable = context.resources.getDrawable(R.drawable.drag_seek_progress)
        drag_iv_block!!.visibility = View.VISIBLE
        drag_iv_cover!!.setImageBitmap(cover)
        tips2ShowAnime(false)
        timeTemp = System.currentTimeMillis()
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        timeUse = (System.currentTimeMillis() - timeTemp) / 1000f
        if (dragListenner != null)
            dragListenner!!.onDrag(
                px2dip(
                    context,
                    (drag_iv_cover!!.measuredWidth - drag_iv_block!!.measuredWidth).toFloat() * 1f * seekBar.progress.toFloat() / seekBar.max
                ).toDouble()
            )
    }
    //===================seekbar监听===================


    //闪烁滑块
    private fun twinkleImage(view: View?) {
        val animator = ValueAnimator.ofFloat(0.0f, 1.0f)
        animator.setTarget(view)
        animator.interpolator = LinearInterpolator()
        animator.setDuration(showTipsTime.toLong()).start()
        animator.addUpdateListener { animation ->
            val f = animation.animatedValue as Float
            val time = (showTipsTime * f).toInt()

            if (time < 125)
                view!!.visibility = View.INVISIBLE
            else if (time < 250)
                view!!.visibility = View.VISIBLE
            else if (time < 375)
                view!!.visibility = View.INVISIBLE
            else
                view!!.visibility = View.VISIBLE
        }
    }

    //提示文本显示隐藏
    private fun tipsShowAnime(isShow: Boolean) {
        if (drag_tv_tips!!.visibility == View.VISIBLE == isShow)
            return
        val translateAnimation = TranslateAnimation(
            Animation.RELATIVE_TO_SELF, 0f,
            Animation.RELATIVE_TO_SELF, 0f,
            Animation.RELATIVE_TO_SELF, if (isShow) 1f else 0f,
            Animation.RELATIVE_TO_SELF, if (isShow) 0f else 1f
        )
        translateAnimation.duration = animeTime.toLong()
        drag_tv_tips!!.animation = translateAnimation
        drag_tv_tips!!.visibility = if (isShow) View.VISIBLE else View.GONE
    }

    //提示文本显示隐藏
    private fun tips2ShowAnime(isShow: Boolean) {
        if (drag_tv_tips2!!.visibility == View.VISIBLE == isShow)
            return
        val translateAnimation =
            AlphaAnimation((if (isShow) 0 else 1).toFloat(), (if (isShow) 1 else 0).toFloat())
        translateAnimation.duration = animeTime.toLong()
        drag_tv_tips2!!.animation = translateAnimation
        drag_tv_tips2!!.visibility = if (isShow) View.VISIBLE else View.GONE
    }

    //成功完成拼图滑块消失
    private fun blockHideAnime() {
        val translateAnimation = AlphaAnimation(1f, 0f)
        translateAnimation.duration = animeTime.toLong()
        drag_iv_block!!.animation = translateAnimation
        drag_iv_block!!.visibility = View.GONE
    }


    //成功高亮动画
    private fun flashShowAnime() {
        val translateAnimation = TranslateAnimation(
            Animation.RELATIVE_TO_SELF, 1f,
            Animation.RELATIVE_TO_SELF, -1f,
            Animation.RELATIVE_TO_SELF, 0f,
            Animation.RELATIVE_TO_SELF, 0f
        )
        translateAnimation.duration = flashTime.toLong()
        drag_v_flash!!.animation = translateAnimation
        drag_v_flash!!.visibility = View.VISIBLE
        translateAnimation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {

            }

            override fun onAnimationEnd(animation: Animation) {
                drag_v_flash!!.visibility = View.GONE
            }

            override fun onAnimationRepeat(animation: Animation) {

            }
        })
    }

    //设置seekBar的Thumb
    fun setSbThumb(id: Int) {
        val drawable = resources.getDrawable(id)
        drawable.bounds = sb!!.thumb.bounds
        sb!!.thumb = drawable
        sb!!.thumbOffset = 0
    }

    //设置seekBar是否可以滑动
    fun setSBUnMove(isMove: Boolean) {
        sb!!.isEnabled = isMove
    }


    interface DragListenner {
        fun onDrag(position: Double)
    }

    fun setDragListenner(dragListenner: DragListenner) {
        this.dragListenner = dragListenner
    }
}